lib/SILOptimizer/Mandatory/TFDeabstraction.cpp

```diff
@@ -1817,37 +1810,47 @@ void TFDeabstraction::cleanupDeadInstructions() {
   llvm::PrettyStackTraceFormat
   X("TFDeabstraction::cleanupDeadInstructions");
 
-  for (auto &block : fn) {
-    for (auto instIt = block.begin(), e = block.end(); instIt != e; ) {
-      // Manually move iterator to avoid invalidation if we remove 'inst'.
-      auto *inst = &*instIt++;
+  SmallVector<SILInstruction *, 16> deadInsts;
+  SmallVector<ApplyInst *, 8> arrayAllocUninitInsts;
 
-      // Do a quick job to remove obvious dead instructions.  Don't delete debug
-      // instructions.
-      if (isa<SingleValueInstruction>(inst) &&
-          isInstructionTriviallyDead(inst)) {
-        recursivelyDeleteTriviallyDeadInstructions(inst);
-        continue;
-      }
+  auto markInstructionForRemoval = [&](SILInstruction *inst) {
+    // Mark trivially dead instructions.
+    if (isa<SingleValueInstruction>(inst) &&
+        isInstructionTriviallyDead(inst)) {
+      deadInsts.push_back(inst);
+      return;
+    }
 
-      // The original call took the array at +0, so we don't need to destroy it.
-      // however, we want to clean up the mess to make the resultant IR simpler.
-      // Do so now if possible.
-      if (auto *apply = dyn_cast<ApplyInst>(inst)) {
-        removeArrayValueIfPossible(apply, instIt);
-        continue;
+    // Mark instructions applying the _allocateUninitialized function.
+    if (auto *apply = dyn_cast<ApplyInst>(inst)) {
+      SILValue tmp;
+      if (isArrayAllocUninit(apply, tmp)) {
+        arrayAllocUninitInsts.push_back(apply);
+        return;
       }
+    }
 
-      // TODO: Remove dead stack slots for tensors being passed as input lists.
+    // TODO: Mark dead stack slots for tensors being passed as input lists.
 
-      // TODO: need to delete TensorShape.init and other methods that we
-      // shouldn't bake in special knowledge of.  Needs integration with
-      // ConstExpr.
+    // TODO: Mark TensorShape.init and other methods that we shouldn't bake in
+    // special knowledge of. Needs integration with ConstExpr.
 
-      // TODO: Handle String.init and other stuff.  Refactor this when ConstExpr
-      // handling subsumes ConstantPropagation.
-    }
-  }
+    // TODO: Handle String.init and other stuff. Refactor this when ConstExpr
+    // handling subsumes ConstantPropagation.
+  };
+
+  // Mark instructions for removal.
+  for (auto &bb : fn)
+    for (auto &inst : bb)
+      markInstructionForRemoval(&inst);
+
+  // Remove dead instructions. Debug instructions are not deleted.
+  for (auto *inst : deadInsts)
+    recursivelyDeleteTriviallyDeadInstructions(inst);
```

Comment:
Is it possible that a previous iteration removed some instruction that's going to be visited in the next iteration? It might not be the case because `markInstructionForRemoval` checks `isInstructionTriviallyDead` so there's no user of instructions visited in the previous iteration.

Reply:
Good point.
For trivially dead instructions, it shouldn't be possible for a previous iteration to delete instructions still in the worklist.
For `_allocateUninitialized` apply instructions, however, this may be possible. The deleted instructions become dangling pointers, which is problematic.

I'm not sure about a general way to make this safe. Currently, each pair of marking/removal functions needs to be consistent to avoid producing dangling pointers.